readme.txt
=== Content Aware Sidebars - Unlimited Widget Areas ===
Contributors: intoxstudio, devinstitute, freemius
Donate link: #
Tags: custom sidebars, page sidebar, replace sidebar, create sidebars, create widget area, bbpress, buddypress, polylang, pods, sidebar manager, sidebar
Requires at least: 4.0
Tested up to: 4.8
Stable tag: 3.6.1
License: GPLv3

Display custom sidebars on any post, page, category etc. Supports bbPress, BuddyPress, WooCommerce, Easy Digital Downloads and more.

== Description ==

**Display tailored sidebars and widget areas on any post, page or custom post type on your WordPress site. No coding required!**

Use Content Aware Sidebars to boost on-site SEO, upsell your products or get better-converting Call-to-Actions by displaying different sidebars on different conditions. The Sidebar Manager makes it incredibly easy to create and manage widget areas in any theme.

The Content Aware Sidebars plugin is built to scale and will never slow down your site no matter the size.

####Create Unlimited Sidebars and Widget Areas

* All-in-one Sidebar Manager
* Create or select sidebars directly when editing a post or page
* Merge with, replace and hide sidebars in any theme
* Activate and deactivate sidebars on schedule
* Sidebar Visibility for All or Logged-in Users
* Enhanced Widget Manager with sidebar and widget search
* Change markup of widget areas
* Optional Template Tag to display custom sidebars anywhere in your theme
* Optional Shortcode to display custom sidebars anywhere in your content
* Multilingual and Translation Ready ([help translate!](https://translate.wordpress.org/projects/wp-plugins/content-aware-sidebars))

####Display Different Sidebars on Any Page

Create a post sidebar, page sidebar, category sidebar, or any custom sidebar you need. Content Aware Sidebars even allows you to combine conditions in any way you like very efficiently. This means that you e.g. can display a custom sidebar on all posts in Category X written by author Y.

For each widget area you create, you can use the following conditions:

* Singulars, eg. each post, page, or custom post type
* Content with select taxonomies, eg. categories or tags
* Content written by a select author
* Page Templates
* Post Type Archives
* Author Archives
* (Custom) Taxonomy Archives
* Date Archives
* Search Results
* 404 Not Found Page
* Front Page
* Blog Page
* bbPress User Profiles
* BuddyPress Member Pages
* Languages (qTranslate X, Polylang, Transposh, WPML)
* Pods Pages

####Sidebar Plugin Integrations and Support

Content Aware Sidebars automatically supports Custom Post Types and Taxonomies created by any plugin or theme. Moreover, it comes with built-in support for some of the most popular WordPress plugins. This means that you e.g. can create WooCommerce sidebars for select products or BuddyPress sidebars for your member sections!

* [bbPress](https://dev.institute/wordpress/sidebars-pro/bbpress/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [BuddyPress](https://dev.institute/wordpress/sidebars-pro/buddypress/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [Easy Digital Downloads](https://dev.institute/wordpress/sidebars-pro/easy-digital-downloads/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [qTranslate X](https://dev.institute/wordpress/sidebars-pro/multilingual-plugins/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [Pods](https://dev.institute/wordpress/sidebars-pro/pods/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [Polylang](https://dev.institute/wordpress/sidebars-pro/multilingual-plugins/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [Transposh Translation Filter](https://dev.institute/wordpress/sidebars-pro/multilingual-plugins/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [WooCommerce](https://dev.institute/wordpress/sidebars-pro/woocommerce/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)
* [WPML](https://dev.institute/wordpress/sidebars-pro/multilingual-plugins/?utm_source=readme&utm_medium=referral&utm_content=integration&utm_campaign=cas)

####Premium Widget Area Manager - Content Aware Sidebars Pro

[Complete control for your custom sidebars](https://dev.institute/wordpress/sidebars-pro/?utm_source=readme&utm_medium=referral&utm_content=title&utm_campaign=cas):

* Priority Email Support
* Automatic Widgets Backup
* Display sidebars on URLs + wildcards
* Display sidebars on content from any day, month, year
* Display sidebars in select time ranges on given days
* Sidebar Visibility for Roles and Guests
* Widget to use sidebars as reusable blocks
* Duplicate Sidebars & Sidebar Conditions
* Disable Widgets
* Sync Sidebar Widgets on Theme Switch
* [and more...](https://dev.institute/wordpress/sidebars-pro/features/?utm_source=readme&utm_medium=referral&utm_content=more&utm_campaign=cas)

You can upgrade at any time securely from your Admin Dashboard via [Freemius](http://freemius.com/)!

####More Information

* [Documentation](https://dev.institute/docs/content-aware-sidebars/?utm_source=readme&utm_medium=referral&utm_content=info&utm_campaign=cas)
* [Github](https://github.com/intoxstudio/content-aware-sidebars)
* [Twitter](https://twitter.com/intoxstudio)

== Installation ==

1. Unzip and upload the `content-aware-sidebars` folder to your `/wp-content/plugins/` directory via FTP or install the plugin in *Plugins* in the Admin Dashboard 
1. Activate the sidebar plugin in *Plugins* in the Admin Dashboard
1. Create your first custom sidebar under the menu *Sidebars > Add New*
1. Add widgets to the new sidebar like any other sidebar
1. That's it!

[Click here to get started with Content Aware Sidebars.](https://dev.institute/docs/content-aware-sidebars/getting-started/?utm_source=readme&utm_medium=referral&utm_content=install&utm_campaign=cas)

* Optional: Insert Template Tag `<?php ca_display_sidebar( $args ); ?>` in your theme for manually handled sidebars
* Optional: Insert Shortcode `[ca-sidebar id=]` in a post or page for manually handled sidebars

== Frequently Asked Questions ==

= Will this sidebar plugin work with my theme? =

Yes!

If your theme already has WordPress sidebars and widget areas, you can create new custom sidebars to replace, merge with or hide those on certain conditions.

If it doesn't, it is still possible to display custom sidebars using the function ca_display_sidebar() in your theme or using the sidebar shortcode [ca-sidebar] in your content.

= Can I really display sidebars and widget areas on any page? =

Yep.

In the Sidebar Editor, you can select on what conditions your sidebar should be displayed. Create a sidebar for each page, a sidebar for posts in a category, or any sidebar you need.
You can also select sidebars directly when you're editing a post or page.

[Click here to read more about sidebar conditions](https://dev.institute/docs/content-aware-sidebars/getting-started/display-sidebar-simple/?utm_source=readme&utm_medium=referral&utm_content=faq&utm_campaign=cas).

= I have other questions, can you help? =

Of course! Check out the links below:

* [Getting Started with Content Aware Sidebars](https://dev.institute/docs/content-aware-sidebars/getting-started/?utm_source=readme&utm_medium=referral&utm_content=faq&utm_campaign=cas)
* [Documentation and FAQ](https://dev.institute/docs/content-aware-sidebars/?utm_source=readme&utm_medium=referral&utm_content=faq&utm_campaign=cas)
* [Support Forums](https://wordpress.org/support/plugin/content-aware-sidebars)

== Screenshots ==

1. Easily create custom sidebars and display them on any content
1. See your widget areas at a glance in the easy Sidebar Manager
1. Filter widgets and sidebars in the Enhanced Widget Manager

== Upgrade Notice ==

= 3.5.1 = 

* Content Aware Sidebars data in your database will be updated automatically. It is highly recommended to backup this data before updating the plugin.
* Data from version 0.8 and below will not be updated during this process.

== Changelog ==

[Follow development on GitHub](https://github.com/intoxstudio/content-aware-sidebars)

= 3.6.1 =

[View Release Notes](https://dev.institute/style-widget-areas-content-aware-sidebars-3-6/?utm_source=readme&utm_medium=referral&utm_content=changelog&utm_campaign=cas)

* Fixed: UI improvements
* Fixed: better plugin compatibility for sidebars quick select

**Pro Plan:**

* Fixed: unable to disable widgets
* Fixed: javascript would in rare cases not be loaded correctly in sidebar editor

= 3.6 =

* Added: ability to wrap sidebar in html
* Added: ability to change widget html
* Added: danish and german translations now use automatic language packs
* Added: performance and memory improvements
* Added: UI improvements
* Added: wordpress 4.8 support
* Fixed: widget manager could behave unexpectedly in cases with many inactive widget areas
* Fixed: sidebars quick select could cause nested sidebars to be removed from posts
* Fixed: sidebar bulk messages
* Updated: wp-content-aware-engine
* Updated: freemius sdk

**Pro Plan:**

* Added: widget to insert widgets from other sidebars
* Added: 1-click automatic install on checkout
* Fixed: avoid auto-draft sidebar adding extra revision 

= 3.5.2 =

[View Release Notes](https://dev.institute/content-aware-sidebars-version-3-5-released/?utm_source=readme&utm_medium=referral&utm_content=changelog&utm_campaign=cas)

* Added: woocommerce 3.0 compat
* Added: badgeos compat
* Added: UI improvements in widget manager

= 3.5.1 =

* Fixed: database upgrade script

= 3.5 =

* Added: 1-click sidebar activation
* Added: improved design on widgets screen
* Added: autosave conditions
* Added: simplify option to autoselect conditions
* Added: UI improvements
* Added: minimum requirement wordpress 4.0
* Added: freemius sdk updated
* Added: actions and filters for developers

**Pro Plan:**

* Added: duplicate sidebars and conditions
* Added: disable widgets
* Fixed: improved upgrade routine

= 3.4.3 =

* Added: preparation for automatic translation packages
* Added: prevent adding duplicate sidebar titles
* Fixed: sidebar quick select would in rare cases not show sidebars
* Fixed: UI improvements

= 3.4.2 =

* Added: freemius sdk updated
* Fixed: sidebar editor now works properly in IE browser
* Fixed: UI improvements

= 3.4.1 =

* Added: ability to target all buddypress profile sections
* Added: freemius opt-in message made more clear
* Added: links to docs
* Added: wordpress 4.7 support
* Fixed: sidebar order not being saved
* Fixed: "Automatically add new children of a selected ancestor" not working for post types
* Fixed: sidebar edit links on widgets screen

**Pro Plan:**

* Fixed: buddypress groups condition not selectable on new sidebars

= 3.4 =

* Added: sidebar list and editor screens completely rewritten for performance and extensibility
* Added: ability to schedule sidebar deactivation
* Added: exposure moved to condition groups, now called singulars or archives
* Added: freemius sdk updated
* Added: data update process will no longer be triggered on new installs
* Added: sidebar status now active/inactive instead of publish/draft
* Added: always load latest version of wp-content-aware-engine
* Fixed: sidebar quick select compatibility with other sidebar managers
* Removed: deprecated function display_ca_sidebar (use ca_display_sidebar)

**Pro Plan:**

* Added: display sidebars in time ranges on given days
* Fixed: initial widget revision could in some cases be malformed
* Fixed: bug when adding url and date conditions
* Fixed: bug with license activation if user had opted out of freemius

= 3.3.3 =

* Added: counter-measure against plugins that add buggy scripts
* Fixed: saving sidebars in quick select would in some cases trigger warning

**Pro Plan:**

* Fixed: include draft sidebars when syncing widgets across themes

= 3.3.2 =

* Fixed: markup in quick select could in some cases be malformed
* Fixed: warning when saving sidebar

= 3.3.1 =

* Added: ux design improvements
* Added: ability to add more sidebars in quick select
* Added: toggle to display more than 3 sidebar input fields in quick select
* Added: select2 dropdowns updated to 4.0.3 on sidebar edit screen
* Added: re-enabled info box on sidebar edit screen
* Added: freemius sdk updated
* Fixed: decoding of taxonomy term names in conditions
* Fixed: order of content in conditions dropdowns
* Fixed: yoast seo compatibility on post edit screens
* Fixed: negated post conditions were included in sidebar quick select
* Removed: upgrade box on sidebar edit screen

**Pro Plan:**

* Fixed: improved widget revision ux
* Fixed: select dates would in some cases not be displayed correct in conditions
* Fixed: widget revisions could in some cases contain wrong or malformed data
* Fixed: backwards compat in widget revision ui for versions before wp4.5

= 3.3 =

* Added: manage widgets for draft sidebars
* Added: quick select and create sidebars on post type editor screen
* Added: ability to add widgets to draft sidebars
* Added: expand/collapse all sidebars on widgets screen
* Added: view sidebar status on widgets screen
* Added: order sidebars by title on widgets screen
* Added: dialog on unsaved condition changes in sidebar editor
* Added: ux design improvements
* Added: more focus on pro features, upgrade box moved to bottom
* Fixed: error in wpml config (props Chouby)
* Fixed: updated review notice description
* Fixed: select2 dropdowns styling more robust to external changes
* Fixed: minor performance improvements

**Pro Plan:**

* Added: set sidebars to published or draft on widgets screen

= 3.2.4 =

* Added: infinite scroll for content in sidebar editor
* Added: guard when activating both free and pro version, uninstall cleanup not run if one is active
* Added: support for buddypress 2.6 members
* Added: wordpress 4.6 support
* Fixed: option to select all authors and bbpress profiles
* Fixed: simplified introduction tour
* Fixed: uninstall cleanup for users not on freemius

**Pro Plan:**

* Fixed: load buddypress group module correctly
* Fixed: search for buddypress groups

= 3.2.3 =

* Fixed: wp function is_user_logged_in would in some cases not be defined in time

= 3.2.2 =

* Fixed: pages with no custom sidebars could cause malformed sql query

= 3.2.1 =

changelog.txt
= 3.1.2 =

* Added: wordpress 4.5 support
* Added: module javascript more extensible
* Fixed: styling for autocomplete input fields
* Fixed: updated translations

= 3.1.1 =

* Added: pods pages module, props @sc0ttkclark @herold
* Fixed: better compat when other themes or plugins load breaking scripts

= 3.1 =

* Added: completely rewritten sidebar editor ui
* Added: refactored plugin into more logical abstractions
* Added: wp-db-updater now handles database updates
* Added: tour reflects new sidebar editor ui
* Added: review notice after 2 weeks use
* Added: admin footer text on relevant pages
* Added: minified some scripts
* Added: qtranslate x module
* Fixed: sidebar and widget filtering for wp4.4+
* Fixed: bug making attachments not selectable
* Fixed: bumped versions for integrated plugins
* Removed: qtranslate module

= 3.0.1 =

* Added: wp4.4 compatibility

= 3.0 =

* Added: plugin rewritten with wp-content-aware-engine to take care of everything related to conditional content
* Added: migrate existing sidebar data to wp-content-aware-engine standards
* Added: negation of condition groups
* Added: new api to add modules
* Added: new action and filter hooks
* Added: intro tour
* Added: ca_display_sidebar deprecates display_ca_sidebar
* Added: filter for sidebars on widgets screen
* Added: add sidebar from widgets screen
* Added: filter for available widgets on widgets screen
* Added: more simple information meta box
* Added: gui improvements in sidebar editor
* Added: bulk update messages translatable
* Added: draft post status included in post type lists
* Added: arabic, bulgarian, croatian, czech, estonian, french (canada), greek, hebrew, indonesian, japanese, persian, vietnamese translations
* Fixed: uninstall script cleans residual data better
* Fixed: posts page and front page excluded from page post type list
* Fixed: updated translations
* Fixed: finnish translation included properly
* Fixed: minimum requirement wp3.8
* Removed: all action and filter hooks (moved to wp-content-aware-engine under new names)
* Removed: deprecated icons
* Removed: screenshots

= 2.6.3 =

* Fixed: sidebars would not be displayed on front page

= 2.6.2 =
* Added: wp4.3 compatibility
* Fixed: sidebars would not be displayed on 404 or empty content
* Fixed: sidebars would not be displayed when queried object differed from post object
* Fixed: included ukranian translation properly

= 2.6.1 =

* Added: finnish, korean, portuguese (portugal), slovenian and swedish translations
* Fixed: updated translations
* Fixed: better compatibility when used together with other sidebar managers

= 2.6 =

* Added: date module to display sidebars on date archives
* Added: basque and french translations
* Fixed: support for mariadb 10.0

= 2.5.3 =

* Added: catalan, norwegian bokmål and turkish translations
* Fixed: updated translations

= 2.5.2 =

* Added: delete_posts capability. fixes conflict in wp4.1
* Fixed: only get custom sidebars when $post is set. fixes conflict in wp4.1

= 2.5.1 =

* Fixed: properly include casmodule.php

= 2.5 =

* Added: ca-sidebar shortcode
* Added: serbian and dutch translations
* Added: teaser meta box
* Added: shortcode is displayed when selecting manual handle in sidebar editor
* Added: description can be added to modules
* Added: generic module callback for ajax allowing search and pagination to use same request
* Added: or-separator between condition groups in editor
* Fixed: headers already sent error on ajax request

= 2.4.3 =

* Fixed: german translation properly included

= 2.4.2 =

* Added: polish, brazilian portuguese and russian translations
* Fixed: "auto-select new children of selected ancestors" caused new top level post types to be selected
* Fixed: "auto-select new children of selected ancestors" for taxonomies could prevent sidebar from being displayed

= 2.4.1 =

* Fixed: authors found via search in sidebar editor could not be saved
* Fixed: displaying max 20 authors in search results instead of 10
* Fixed: improved ux design for sidebars in widgets screen

= 2.4 =

* Added: compatibility with wp4.0
* Added: better ux design for condition groups
* Added: better ux design for sidebars in widgets screen, including an edit link
* Added: dashicon for admin menu (wp3.8+)
* Fixed: using some newer wordpress actions and filters for admin columns
* Fixed: sidebars could be fetched and prepared in administration

= 2.3 =

* Added: ukranian translation
* Added: error if trying to access php files directly
* Added: bbpress, bp_member, polylang, qtranslate, transposh and wpml modules are more robust
* Added: content rule boxes can be hidden in screen options in sidebar editor
* Added: help tab in sidebar editor
* Added: widgets count column in sidebar overview
* Added: menu order moved to options meta box in sidebar editor
* Fixed: merge position option hidden on forced replace handle
* Fixed: polylang compatibility now 1.2+
* Fixed: width of columns in sidebar overview
* Removed: exposure column in sidebar overview

= 2.2.1 =

* Fixed: taxonomy archive conditions did not work with other modules properly
* Fixed: removed display limit of 20 for saved post type conditions
* Fixed: saved post type conditions ordered alphabetically

= 2.2 =

* Added: sidebar displayed in theme customizer (wp3.4+)
* Added: widget management integration in theme customizer (wp3.9+)
* Added: handle for forced replace
* Fixed: reduced database queries in sidebar editor
* Fixed: disable all add to group buttons before creating first condition group

= 2.1 =

* Added: empty condition groups cannot be saved
* Added: confirmation on various condition group actions in sidebar editor
* Added: improved ux design for sidebar editor
* Added: chinese translation
* Added: wp3.9 compatibility
* Fixed: transposh compatibility now 0.9.5+

sidebar.php
<?php
/**
 * @package Content Aware Sidebars
 * @author Joachim Jensen <jv@intox.dk>
 * @license GPLv3
 * @copyright 2017 by Joachim Jensen
 */

if (!defined('ABSPATH')) {
	exit;
}

final class CAS_Sidebar_Manager {

	/**
	 * Sidebar metadata
	 * @var WPCAObjectManager
	 */
	protected $metadata;

	/**
	 * Custom sidebars
	 * @var array
	 */
	public $sidebars = array();

	/**
	 * Cache replaced sidebars
	 * @var array
	 */
	protected $replaced_sidebars = array();

	/**
	 * Sidebar replacement map
	 * @var array
	 */
	protected $replace_map = array();

	/**
	 * @var array
	 * Constructor
	 *
	 * @since 3.1
	 */
	public function __construct() {

		add_action('wpca/loaded',
			array($this,'late_init'));
		add_action('wp_head',
			array($this,'sidebar_notify_theme_customizer'));
		add_action('init',
			array($this,'init_sidebar_type'),99);
		add_action('widgets_init',
			array($this,'create_sidebars'),99);
		add_action('wp_loaded',
			array($this,'set_sidebar_styles'),99);

		add_shortcode( 'ca-sidebar',
			array($this,'sidebar_shortcode'));


	}

	/**
	 * Initialize after WPCA has been loaded
	 * Makes sure the SDK can be used in actions/filters
	 * forcefully called earlier
	 *
	 * @since  3.4
	 * @return void
	 */
	public function late_init() {
		if(!is_admin()) {
			add_filter('sidebars_widgets',
				array($this,'replace_sidebar'));
			add_action( 'dynamic_sidebar_before',
				array($this,'render_sidebar_before'),9,2);
			add_action( 'dynamic_sidebar_after',
				array($this,'render_sidebar_after'),99,2);
		}
	}

	/**
	 * Get instance of metadata manager
	 *
	 * @since  3.0
	 * @return WPCAObjectManager
	 */
	public function metadata() {
		if(!$this->metadata) {
			$this->init_metadata();
		}
		return $this->metadata;
	}

	/**
	 * Create post meta fields
	 * @global array $wp_registered_sidebars 
	 * @return void 
	 */
	private function init_metadata() {

		$this->metadata = new WPCAObjectManager();
		$this->metadata
		->add(new WPCAMeta(
			'visibility',
			__('Visibility','content-aware-sidebars'),
			array(),
			'multi',
			array(
				-1 => __('Logged-in Users', 'content-aware-sidebars')
			)
		),'visibility')
		->add(new WPCAMeta(
			'handle',
			_x('Action','option', 'content-aware-sidebars'),
			0,
			'select',
			array(
				0 => __('Replace', 'content-aware-sidebars'),
				1 => __('Merge', 'content-aware-sidebars'),
				3 => __('Forced replace','content-aware-sidebars'),
				2 => __('Shortcode / Template Tag', 'content-aware-sidebars')
			),
			__('Replace host sidebar, merge with it or add sidebar manually.', 'content-aware-sidebars')
		),'handle')
		->add(new WPCAMeta(
			'host',
			__('Target Sidebar', 'content-aware-sidebars'),
			'sidebar-1',
			'select',
			array()
		),'host')
		->add(new WPCAMeta(
			'merge_pos',
			__('Merge Position', 'content-aware-sidebars'),
			1,
			'select',
			array(
				__('Top', 'content-aware-sidebars'),
				__('Bottom', 'content-aware-sidebars')
			),
			__('Place sidebar on top or bottom of host when merging.', 'content-aware-sidebars')
		),'merge_pos')
		->add(new WPCAMeta(
			'html',
			__('HTML', 'content-aware-sidebars'),
			array(),
			'select',
			array('')
		),'html');
		apply_filters('cas/metadata/init',$this->metadata);
	}

	/**
	 * Populate metadata with dynamic content
	 * for use in admin
	 *
	 * @since  3.2
	 * @return void
	 */
	public function populate_metadata() {
		if($this->metadata) {

			global $wp_registered_sidebars;

			// List of sidebars
			$sidebar_list = array();
			foreach($wp_registered_sidebars as $sidebar) {
				$sidebar_list[$sidebar['id']] = $sidebar['name'];
			}

			// Remove ability to set self to host
			if(get_the_ID()) {
				unset($sidebar_list[CAS_App::SIDEBAR_PREFIX.get_the_ID()]);
			}
			$this->metadata->get('host')->set_input_list($sidebar_list);

			apply_filters('cas/metadata/populate',$this->metadata);
		}
		
	}

	/**
	 * Create sidebar post type
	 * Add it to content aware engine
	 * 
	 * @return void 
	 */
	public function init_sidebar_type() {
		
		// Register the sidebar type
		register_post_type(CAS_App::TYPE_SIDEBAR,array(
			'labels'        => array(
				'name'               => __('Sidebars', 'content-aware-sidebars'),
				'singular_name'      => __('Sidebar', 'content-aware-sidebars'),
				'add_new'            => _x('Add New', 'sidebar', 'content-aware-sidebars'),
				'add_new_item'       => __('Add New Sidebar', 'content-aware-sidebars'),
				'edit_item'          => __('Edit Sidebar', 'content-aware-sidebars'),
				'new_item'           => __('New Sidebar', 'content-aware-sidebars'),
				'all_items'          => __('All Sidebars', 'content-aware-sidebars'),
				'view_item'          => __('View Sidebar', 'content-aware-sidebars'),
				'search_items'       => __('Search Sidebars', 'content-aware-sidebars'),
				'not_found'          => __('No sidebars found', 'content-aware-sidebars'),
				'not_found_in_trash' => __('No sidebars found in Trash', 'content-aware-sidebars'),
				//wp-content-aware-engine specific
				'ca_title'           => __('Sidebar Conditions','content-aware-sidebars'),
				'ca_not_found'       => __('No content. Please add at least one condition group to make the sidebar content aware.','content-aware-sidebars')
			),
			'capabilities'  => array(
				'edit_post'          => CAS_App::CAPABILITY,
				'read_post'          => CAS_App::CAPABILITY,
				'delete_post'        => CAS_App::CAPABILITY,
				'edit_posts'         => CAS_App::CAPABILITY,
				'delete_posts'       => CAS_App::CAPABILITY,
				'edit_others_posts'  => CAS_App::CAPABILITY,
				'publish_posts'      => CAS_App::CAPABILITY,
				'read_private_posts' => CAS_App::CAPABILITY
			),
			'public'              => false,
			'hierarchical'        => false,
			'exclude_from_search' => true,
			'publicly_queryable'  => false,
			'show_ui'             => false,
			'show_in_menu'        => false,
			'show_in_nav_menus'   => false,
			'show_in_admin_bar'   => false,
			'has_archive'         => false,
			'rewrite'             => false,
			'query_var'           => false,
			'supports'            => array('title','page-attributes'),
			'menu_icon'           => 'dashicons-welcome-widgets-menus',
			'can_export'          => false,
			'delete_with_user'    => false
		));

		WPCACore::post_types()->add(CAS_App::TYPE_SIDEBAR);
	}

	/**
	 * Add sidebars to widgets area
	 * Triggered in widgets_init to save location for each theme
	 * @return void
	 */
	public function create_sidebars() {
		$sidebars = get_posts(array(
			'numberposts' => -1,
			'post_type'   => CAS_App::TYPE_SIDEBAR,
			'post_status' => array(
				CAS_App::STATUS_ACTIVE,
				CAS_App::STATUS_INACTIVE,
				CAS_App::STATUS_SCHEDULED
			),
			'orderby'     => 'title',
			'order'       => 'ASC'
		));

		//Register sidebars to add them to the list
		foreach($sidebars as $post) {
			$this->sidebars[CAS_App::SIDEBAR_PREFIX.$post->ID] = $post;
			register_sidebar( array(
				'name'           => $post->post_title ? $post->post_title : __('(no title)'),
				'id'             => CAS_App::SIDEBAR_PREFIX.$post->ID,
				'before_sidebar' => '',
				'after_sidebar'  => ''
			));
		}
	}
	
	/**
	 * Set styles of created sidebars
	 *
	 * @since 3.6
	 */
	public function set_sidebar_styles() {
		global $wp_registered_sidebars;

		//todo: only for manual
		$default_styles = array(
			'before_widget' => '<li id="%1$s" class="widget-container %2$s">',
			'after_widget'  => '</li>',
			'before_title'  => '<h4 class="widget-title">',
			'after_title'   => '</h4>'
		);
		$has_host = array(0=>1,1=>1,3=>1);
		$metadata = $this->metadata();

		foreach($this->sidebars as $post) {
			$id = CAS_App::SIDEBAR_PREFIX.$post->ID;

			$args = $default_styles;

			if (isset($has_host[$metadata->get('handle')->get_data($post->ID)])) {
				//Set style from host to fix when content aware sidebar
				//is called directly by other sidebar managers
				$host_id = $metadata->get('host')->get_data($post->ID);
				if(isset($wp_registered_sidebars[$host_id])) {
					foreach (array(
						'before_widget',
						'after_widget',
						'before_title',
						'after_title',
						'before_sidebar',
						'after_sidebar'
					) as $pos) {
						if(isset($wp_registered_sidebars[$host_id][$pos])) {
							$args[$pos] = $wp_registered_sidebars[$host_id][$pos];
						}
					}
				}
			}

			$wp_registered_sidebars[$id] = array_merge($wp_registered_sidebars[$id],$args);
		}
	}

	/**
	 * Replace or merge a sidebar with content aware sidebars.
	 * @since  .
	 * @param  array    $sidebars_widgets
	 * @return array
	 */
	public function replace_sidebar($sidebars_widgets) {

		//customizer requires sidebars_widgets filter. cache for repeat calls
		if($this->replaced_sidebars) {
			return $this->replaced_sidebars;
		}

		$posts = WPCACore::get_posts(CAS_App::TYPE_SIDEBAR);

		if ($posts) {
			global $wp_registered_sidebars;

			$metadata = $this->metadata();

			//temporary filter until WPCACore allows filtering
			$user_visibility = is_user_logged_in() ? array(-1) : array();
			$user_visibility = apply_filters('cas/user_visibility',$user_visibility);
			foreach ($posts as $post) {

				$id = CAS_App::SIDEBAR_PREFIX . $post->ID;
				$visibility = $metadata->get('visibility')->get_data($post->ID,true,false);
				$host = $metadata->get('host')->get_data($post->ID);

				// Check visibility
				if($visibility && !array_intersect($visibility,$user_visibility)) {
					continue;
				}

				// Check for correct handling and if host exist
				if ( $post->handle == 2 || !isset($sidebars_widgets[$host])) {
					continue;
				}

				// Sidebar might not have any widgets. Get it anyway!
				if (!isset($sidebars_widgets[$id])) {
					$sidebars_widgets[$id] = array();
				}

				// If handle is merge or if handle is replace and host has already been replaced
				if ($post->handle == 1 || ($post->handle == 0 && isset($handled_already[$host]))) {
					if ($metadata->get('merge_pos')->get_data($post->ID)) {
						$sidebars_widgets[$host] = array_merge($sidebars_widgets[$host], $sidebars_widgets[$id]);
					} else {
						$sidebars_widgets[$host] = array_merge($sidebars_widgets[$id], $sidebars_widgets[$host]);
					}
				} else {
					$sidebars_widgets[$host] = $sidebars_widgets[$id];
					$handled_already[$host] = 1;
				}

				//last replacement will take priority
				//todo: extend to work for widgets too
				$this->replace_map[$host] = $id;
			}
			$this->replaced_sidebars = $sidebars_widgets;
		}
		return $sidebars_widgets;
	}

	/**
	 * Show manually handled content aware sidebars
	 * @global array $_wp_sidebars_widgets
	 * @param  string|array $args 
	 * @return void 
	 */
	public function manual_sidebar($args) {
		global $_wp_sidebars_widgets;

		// Grab args or defaults
		$args = wp_parse_args($args, array(
			'include' => '',
			'before'  => '',
			'after'   => ''
		));
		extract($args, EXTR_SKIP);

		// Get sidebars
		$posts = WPCACore::get_posts(CAS_App::TYPE_SIDEBAR);
		if (!$posts)
			return;

		// Handle include argument
		if (!empty($include)) {
			if (!is_array($include))
				$include = explode(',', $include);
			// Fast lookup
			$include = array_flip($include);
		}

		$i = $host = 0;
		foreach ($posts as $post) {

			$id = CAS_App::SIDEBAR_PREFIX . $post->ID;

			// Check for manual handling, if sidebar exists and if id should be included
			if ($post->handle != 2 || !isset($_wp_sidebars_widgets[$id]) || (!empty($include) && !isset($include[$post->ID])))
				continue;

			// Merge if more than one. First one is host.
			if ($i > 0) {
				if ($this->metadata()->get('merge_pos')->get_data($post->ID))
					$_wp_sidebars_widgets[$host] = array_merge($_wp_sidebars_widgets[$host], $_wp_sidebars_widgets[$id]);
				else
					$_wp_sidebars_widgets[$host] = array_merge($_wp_sidebars_widgets[$id], $_wp_sidebars_widgets[$host]);
			} else {
				$host = $id;
			}
			$i++;
		}

		if ($host) {
			echo $before;
			dynamic_sidebar($host);
			echo $after;
		}
	}

	/**
	 * Display sidebar with shortcode
	 * @version 2.5
	 * @param   array     $atts
	 * @param   string    $content
	 * @return  string
	 */
	public function sidebar_shortcode( $atts, $content = '' ) {
		$a = shortcode_atts( array(
			'id' => 0,
		), $atts );
		
		$id = CAS_App::SIDEBAR_PREFIX.esc_attr($a['id']);

		//if sidebar is in replacement map, shortcode is called wrongly
		if(!isset($this->replace_map[$id])) {
			ob_start();
			dynamic_sidebar($id);
			$content = ob_get_clean();
		}
		return $content;
	}

	/**
	 * Get styles from nested sidebars
	 *
	 * @since  3.6
	 * @param  string  $i
	 * @param  array   $styles
	 * @return array
	 */
	public function get_sidebar_styles($i,$styles) {
		if(isset($this->replace_map[$i])) {
			$styles = $this->get_sidebar_styles($this->replace_map[$i],$styles);
		}

		if(isset($this->sidebars[$i])) {
			$html = $this->metadata()->get('html')->get_data($this->sidebars[$i]->ID);
			//Set user styles
			foreach (array(
				'widget',
				'title',
				'sidebar'
			) as $pos) {
				if(isset($html[$pos],$html[$pos.'_class'])) {
					$e = esc_html($html[$pos]);
					$class = esc_html($html[$pos.'_class']);
					$id = '';
					if($pos == 'widget') {
						$id = ' id="%1$s"';
					}
					$styles['before_'.$pos] = '<'.$e.$id.' class="'.$class.'">';
					$styles['after_'.$pos] = "</$e>";
				}
			}
		}

		return $styles;
	}

	/**
	 * Render html if present before sidebar
	 *
	 * @since  3.6
	 * @param  string   $i
	 * @param  boolean  $has_widgets
	 * @return void
	 */
	public function render_sidebar_before($i,$has_widgets) {
		global $wp_registered_sidebars;

		//Get nested styles
		$wp_registered_sidebars[$i] = $this->get_sidebar_styles($i,$wp_registered_sidebars[$i]);

		if($has_widgets && isset($wp_registered_sidebars[$i]['before_sidebar'])) {
			echo $wp_registered_sidebars[$i]['before_sidebar'];
		}
	}

	/**
	 * Render html if present after sidebar
	 *
	 * @since  3.6
	 * @param  string   $i
	 * @param  boolean  $has_widgets
	 * @return void
	 */
	public function render_sidebar_after($i,$has_widgets) {
		global $wp_registered_sidebars;
		if($has_widgets && isset($wp_registered_sidebars[$i]['after_sidebar'])) {
			echo $wp_registered_sidebars[$i]['after_sidebar'];
		}
	}

	/**
	 * Runs is_active_sidebar for sidebars
	 * Widget management in Theme Customizer
	 * expects this
	 * 
	 * @global type $wp_customize
	 * @since  2.2
	 * @return void
	 */
	public function sidebar_notify_theme_customizer() {
		global $wp_customize;
		if(!empty($wp_customize)) {
			$sidebars = WPCACore::get_posts(CAS_App::TYPE_SIDEBAR);
			if($sidebars) {
				foreach($sidebars as $sidebar) {
					is_active_sidebar(CAS_App::SIDEBAR_PREFIX . $sidebar->ID);
				}
			}
		}
	}

}

//eol